### PR TITLE
feat: add bandersnatch curve

### PR DESCRIPTION
**Pull request by dragan2234**
# Pull Request type

Adding bandersnatch parameters https://eprint.iacr.org/2021/1152.pdf

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: N/A

# What is the new behavior?

-
-
-

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

# Other information

### CODE REVIEW

File: hydra/definitions.py [hunk 1]
```diff
@@ -233,6 +235,26 @@ def NAF(x):
...
+        nr_a0=None,
+        nr_a1=None,
+        a=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFFFFE1FB22001,  # See https://neuromancer.sk/std/other/Curve25519 for weirstrass form conversion utilities
+        b=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFB870D2E00001,
+        b20=None,
+        b21=None,
+        loop_counter=None,
+        line_function_sparsity=None,
+        final_exp_cofactor=None,
+        fp_generator=3, # todo: check if this is true
```
Comment (feltroidprime) [inline hydra/definitions.py:254]: use the function is_generator in this file to check. If not, try 2,3,4,5,6,7, etc... you will find one

File: src/fustat/definitions.cairo [hunk 2]
```diff
@@ -118,6 +118,52 @@ namespace secp256r1 {
     const MIN_ONE_D3 = 0x0;
 }
 
+namespace bandersnatch {
```
Comment (feltroidprime) [inline src/fustat/definitions.cairo:121]: generate this code automatically with the https://github.com/keep-starknet-strange/garaga/blob/cd329cc873b83841b0828080f6e6db95cae87ef8/hydra/definitions.py#L53-L75 to_cairo_zero method

**Review by feltroidprime — CHANGES_REQUESTED** [review]

File: hydra/definitions.py [hunk 3]
```diff
@@ -233,6 +235,26 @@ def NAF(x):
...
+    BANDERSNATCH_ID: Curve(
+        cairo_zero_namespace_name="bandersnatch",
+        id=BANDERSNATCH_ID,
+        p=0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001,
+        n=0x1cfb69d4ca675f520cce760202687600ff8f87007419047174fd06b52876e7e1,
+        h=4,
+        irreducible_polys={},
+        nr_a0=None,
+        nr_a1=None,
+        a=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFFFFE1FB22001,  # See https://neuromancer.sk/std/other/Curve25519 for weirstrass form conversion utilities
```
Comment (feltroidprime) [inline hydra/definitions.py:247]: clean the comment related to curve25519

**Review by feltroidprime — CHANGES_REQUESTED** [review]

File: hydra/definitions.py [hunk 4]
```diff
@@ -233,6 +235,26 @@ def NAF(x):
...
+        b=0x73EDA753299D7D483339D80809A1D80553BDA402FFFE5BFEFFFFB870D2E00001,
```
Comment (feltroidprime) [inline hydra/definitions.py:248]: the a and b parameters seems wrong according to https://eprint.iacr.org/2021/1152.pdf section 3.1

**Review by feltroidprime — CHANGES_REQUESTED** [review]

File: hydra/definitions.py (hunk 4 above)
Comment (dragan2234) [reply hydra/definitions.py:248]: @feltroidprime 

`a` is −3763200000 mod 52435875175126190479447740508185965837690552500527637822603658699934817984513

same is with `b`

Comment (feltroidprime) [reply hydra/definitions.py:248]: right

**Comment by dragan2234** [comment]
@feltroidprime Thanks for the review, I have addressed the comments! It should be fine now. add_ec test is passing, but locally I still have some issues with the setup on msm:
```
Exception: Error executing ecip-cli: python3: can't open file '/tools/ecip/main.sage.py': [Errno 2] No such file or directory
```
when I try executing `make run` `msm_ecip.cairo`

**Comment by feltroidprime** [comment]
> @feltroidprime Thanks for the review, I have addressed the comments! It should be fine now.
> 
> Strange thing is that msm for 1,2,3 points are fine with bandersantch, but 256 points is failing for some reason

Yes it is expected, circuits have been compiled for up to 3 points for the LHS of the equation to verify https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/src/fustat/precompiled_circuits/ec.cairo#L10 

For the RHS, it supports any number of points as it calls the same circuit in a loop : https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/src/fustat/ec_ops.cairo#L233 but for the LHS we would need to modify the outputs / add an extra circuit to call in a loop to allow the LHS computation for arbitrary number of points https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/hydra/precompiled_circuits/ec.py#L243 , depending on the number of coefficients in sum_dlog_div , see number of coefficients in function of n : https://github.com/keep-starknet-strange/garaga/blob/dac43112e3298cf64cc8eb656dff05f1d4072ec6/hydra/precompiled_circuits/all_circuits.py#L352-L357

For now, please just comment the n=256 line in the test so I can merge this, and we can add it later.

**Comment by dragan2234** [comment]
@feltroidprime Thanks for the explanation, just commented out line for testing 256 points

**Pull request closed without merging**